File: backend/app/api/v1/monitor/monitor.py
from flask import Blueprint, jsonify, request
from app.core.security.auth import token_required
from app.models.models import db, SystemMetrics, NetworkMetrics, DiskMetrics, ProcessMetrics
from datetime import datetime, timedelta
from sqlalchemy import func

monitor_bp = Blueprint('monitor', __name__)

@monitor_bp.route('/monitor', methods=['GET'])
@token_required
def get_monitor_data(current_user):
    """获取各资源独立趋势数据"""
    try:
        # 获取时间范围参数
        time_range = request.args.get('time_range', '1h')
        hours = {
            '1h': 1,
            '6h': 6,
            '24h': 24
        }.get(time_range, 1)

        end_time = datetime.utcnow()
        start_time = end_time - timedelta(hours=hours)

        # CPU/内存趋势（SystemMetrics）
        cpu_metrics = SystemMetrics.query.filter(
            SystemMetrics.timestamp >= start_time,
            SystemMetrics.timestamp <= end_time
        ).order_by(SystemMetrics.timestamp.asc()).all()
        cpu_trend = {
            'timestamps': [m.timestamp.strftime('%Y-%m-%d %H:%M:%S') for m in cpu_metrics],
            'cpu': [m.cpu_usage for m in cpu_metrics]
        }
        memory_trend = {
            'timestamps': [m.timestamp.strftime('%Y-%m-%d %H:%M:%S') for m in cpu_metrics],
            'memory': [m.memory_usage for m in cpu_metrics]
        }

        # 磁盘趋势（DiskMetrics）
        disk_metrics = DiskMetrics.query.filter(
            DiskMetrics.timestamp >= start_time,
            DiskMetrics.timestamp <= end_time
        ).order_by(DiskMetrics.timestamp.asc()).all()
        # 按时间点聚合
        disk_trend_map = {}
        for m in disk_metrics:
            ts = m.timestamp.strftime('%Y-%m-%d %H:%M:%S')
            if ts not in disk_trend_map:
                disk_trend_map[ts] = {
                    'read': 0, 'write': 0, 'iops': 0
                }
            disk_trend_map[ts]['read'] += m.read_bytes
            disk_trend_map[ts]['write'] += m.write_bytes
            disk_trend_map[ts]['iops'] += (m.read_count + m.write_count)
        disk_trend = {
            'timestamps': list(disk_trend_map.keys()),
            'disk_read': [v['read'] for v in disk_trend_map.values()],
            'disk_write': [v['write'] for v in disk_trend_map.values()],
            'disk_iops': [v['iops'] for v in disk_trend_map.values()]
        }

        # 网络趋势（NetworkMetrics）
        network_metrics = NetworkMetrics.query.filter(
            NetworkMetrics.timestamp >= start_time,
            NetworkMetrics.timestamp <= end_time
        ).order_by(NetworkMetrics.timestamp.asc()).all()
        network_trend_map = {}
        for m in network_metrics:
            ts = m.timestamp.strftime('%Y-%m-%d %H:%M:%S')
            if ts not in network_trend_map:
                network_trend_map[ts] = {'sent': 0, 'recv': 0}
            network_trend_map[ts]['sent'] += m.bytes_sent
            network_trend_map[ts]['recv'] += m.bytes_recv
        network_trend = {
            'timestamps': list(network_trend_map.keys()),
            'network_sent': [v['sent'] for v in network_trend_map.values()],
            'network_recv': [v['recv'] for v in network_trend_map.values()]
        }

        # 最新快照
        latest_metrics = SystemMetrics.get_latest_metrics()
        latest_disk_time = db.session.query(func.max(DiskMetrics.timestamp)).scalar()
        latest_partitions = DiskMetrics.query.filter(DiskMetrics.timestamp == latest_disk_time).all() if latest_disk_time else []
        latest_network_time = db.session.query(func.max(NetworkMetrics.timestamp)).scalar()
        latest_networks = NetworkMetrics.query.filter(NetworkMetrics.timestamp == latest_network_time).all() if latest_network_time else []
        latest_process_time = db.session.query(func.max(ProcessMetrics.timestamp)).scalar()
        latest_processes = ProcessMetrics.query.filter(ProcessMetrics.timestamp == latest_process_time).order_by(ProcessMetrics.cpu_percent.desc()).limit(50).all() if latest_process_time else []

        response_data = {
            'code': 200,
            'data': {
                'cpu_trend': cpu_trend,
                'memory_trend': memory_trend,
                'disk_trend': disk_trend,
                'network_trend': network_trend,
                'stats': {
                    'cpu_usage': latest_metrics.cpu_usage if latest_metrics else 0,
                    'memory_usage': latest_metrics.memory_usage if latest_metrics else 0,
                    'process_count': latest_metrics.process_count if latest_metrics else 0,
                    'cpu_count': latest_metrics.cpu_count if latest_metrics else 0,
                    'memory_total': latest_metrics.memory_total if latest_metrics else 0,
                    'memory_used': latest_metrics.memory_used if latest_metrics else 0,
                    'memory_free': latest_metrics.memory_free if latest_metrics else 0,
                    'load_avg_1min': latest_metrics.load_avg_1min if latest_metrics else 0,
                    'load_avg_5min': latest_metrics.load_avg_5min if latest_metrics else 0,
                    'load_avg_15min': latest_metrics.load_avg_15min if latest_metrics else 0
                },
                'disk': {
                    'partitions': [
                        {
                            'device': m.device,
                            'mountpoint': m.mountpoint,
                            'total': m.total,
                            'used': m.used,
                            'free': m.free,
                            'usage': m.usage,
                            'read_bytes': m.read_bytes,
                            'write_bytes': m.write_bytes,
                            'read_count': m.read_count,
                            'write_count': m.write_count,
                            'read_time': m.read_time,
                            'write_time': m.write_time,
                            'is_removable': m.is_removable,
                            'fstype': m.fstype
                        }
                        for m in latest_partitions
                    ]
                },
                'network': {
                    'interfaces': [
                        {
                            'name': n.interface,
                            'bytes_sent': n.bytes_sent,
                            'bytes_recv': n.bytes_recv,
                            'packets_sent': n.packets_sent,
                            'packets_recv': n.packets_recv,
                            'errin': n.errin,
                            'errout': n.errout,
                            'dropin': n.dropin,
                            'dropout': n.dropout,
                            'is_up': n.is_up,
                            'speed': n.speed,
                            'mtu': n.mtu
                        }
                        for n in latest_networks
                    ]
                },
                'processes': [
                    {
                        'pid': p.pid,
                        'name': p.name,
                        'cpu_percent': p.cpu_percent,
                        'memory_percent': p.memory_percent,
                        'memory_rss': p.memory_rss,
                        'memory_vms': p.memory_vms,
                        'status': p.status,
                        'num_threads': p.num_threads,
                        'create_time': p.create_time.strftime('%Y-%m-%d %H:%M:%S') if p.create_time else None
                    }
                    for p in latest_processes
                ]
            }
        }
        return jsonify(response_data)
    except Exception as e:
        return jsonify({
            'code': 500,
            'message': f'获取监控数据失败: {str(e)}'
        }), 500 

File: frontend/src/stores/monitor.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import axios from 'axios'
import { API_CONFIG } from '../config/api'

interface SystemStats {
  cpu_usage: number
  memory_usage: number
  process_count: number
  cpu_count: number
  disk_usage: number
  memory_total: number
  memory_used: number
  memory_free: number
  load_avg_1min: number
  load_avg_5min: number
  load_avg_15min: number
}

interface CpuTrend {
  timestamps: string[]
  cpu: number[]
}

interface MemoryTrend {
  timestamps: string[]
  memory: number[]
}

interface DiskTrend {
  timestamps: string[]
  disk_read: number[]
  disk_write: number[]
  disk_iops: number[]
}

interface NetworkTrend {
  timestamps: string[]
  network_sent: number[]
  network_recv: number[]
}

interface Process {
  pid: number
  name: string
  cpu_percent: number
  memory_percent: number
  memory_rss: number
  memory_vms: number
  status: string
  num_threads: number
  create_time: string
}

interface NetworkInterface {
  name: string
  bytes_sent: number
  bytes_recv: number
  packets_sent: number
  packets_recv: number
  errin: number
  errout: number
  dropin: number
  dropout: number
  is_up: boolean
  speed: number
  mtu: number
}

interface DiskPartition {
  device: string
  mountpoint: string
  total: number
  used: number
  free: number
  usage: number
  read_bytes: number
  write_bytes: number
  read_count: number
  write_count: number
  read_time: number
  write_time: number
  is_removable: boolean
  fstype: string
}

export const useMonitorStore = defineStore('monitor', () => {
  const systemStats = ref<SystemStats>({
    cpu_usage: 0,
    memory_usage: 0,
    disk_usage: 0,
    process_count: 0,
    cpu_count: 0,
    memory_total: 0,
    memory_used: 0,
    memory_free: 0,
    load_avg_1min: 0,
    load_avg_5min: 0,
    load_avg_15min: 0
  })

  const cpuTrend = ref<CpuTrend>({ timestamps: [], cpu: [] })
  const memoryTrend = ref<MemoryTrend>({ timestamps: [], memory: [] })
  const diskTrend = ref<DiskTrend>({ timestamps: [], disk_read: [], disk_write: [], disk_iops: [] })
  const networkTrend = ref<NetworkTrend>({ timestamps: [], network_sent: [], network_recv: [] })

  const processes = ref<Process[]>([])
  const timeRange = ref<'1h' | '6h' | '24h'>('1h')
  const loading = ref(false)
  const error = ref<string | null>(null)
  const networkInfo = ref<{ interfaces: NetworkInterface[] }>({ interfaces: [] })
  const diskInfo = ref<{ partitions: DiskPartition[] }>({ partitions: [] })

  let pollInterval: number | null = null

  const fetchMonitorData = async () => {
    try {
      loading.value = true
      error.value = null
      const response = await axios.get(`${API_CONFIG.BASE_API_URL}${API_CONFIG.ENDPOINTS.MONITOR.LIST}`, {
        params: { time_range: timeRange.value }
      })
      if (response.data.code === 200) {
        const data = response.data.data
        cpuTrend.value = data.cpu_trend
        memoryTrend.value = data.memory_trend
        diskTrend.value = data.disk_trend
        networkTrend.value = data.network_trend
        systemStats.value = data.stats
        diskInfo.value = data.disk
        networkInfo.value = data.network
        processes.value = data.processes
      } else {
        error.value = response.data.message || '获取监控数据失败'
      }
    } catch (e) {
      error.value = e instanceof Error ? e.message : '获取监控数据失败'
    } finally {
      loading.value = false
    }
  }

  const setTimeRange = (range: '1h' | '6h' | '24h') => {
    timeRange.value = range
    fetchMonitorData()
  }

  const init = () => {
    fetchMonitorData()
    pollInterval = window.setInterval(fetchMonitorData, 5000)
    return () => {
      if (pollInterval) {
        clearInterval(pollInterval)
      }
    }
  }

  return {
    systemStats,
    cpuTrend,
    memoryTrend,
    diskTrend,
    networkTrend,
    processes,
    timeRange,
    loading,
    error,
    networkInfo,
    diskInfo,
    setTimeRange,
    init
  }
})
